data/transactions/logic/debugger.go

```diff
@@ -16,294 +16,143 @@
+	// groupIndex refers to the index of the transaction in the transaction group that was just executed.
+	AfterTxn(ep *EvalParams, groupIndex int) error
 
-	// initialize DebuggerState with immutable fields
-	ds := &DebugState{
-		ExecID:      GetProgramID(cx.program),
-		Disassembly: disasm,
-		PCOffset:    dsInfo.pcOffset,
-		GroupIndex:  int(cx.groupIndex),
-		TxnGroup:    cx.TxnGroup,
-		Proto:       cx.Proto,
-	}
+	// BeforeLogicEval is called before an app or LogicSig is evaluated.
+	BeforeLogicEval(cx *EvalContext) error
 
-	globals := make([]basics.TealValue, len(globalFieldSpecs))
-	for _, fs := range globalFieldSpecs {
-		// Don't try to grab app only fields when evaluating a signature
-		if (cx.runModeFlags&modeSig) != 0 && fs.mode == modeApp {
-			continue
-		}
-		sv, err := cx.globalFieldToValue(fs)
-		if err != nil {
-			sv = stackValue{Bytes: []byte(err.Error())}
-		}
-		globals[fs.field] = stackValueToTealValue(&sv)
-	}
-	ds.Globals = globals
+	// AfterLogicEval is called after an app or LogicSig is evaluated.
+	AfterLogicEval(cx *EvalContext, evalError error) error
 
-	if (cx.runModeFlags & modeApp) != 0 {
-		ds.EvalDelta = cx.txn.EvalDelta
-	}
+	// BeforeTealOp is called before the op is evaluated
+	BeforeTealOp(cx *EvalContext) error
 
-	return ds
-}
+	// AfterTealOp is called after the op has been evaluated
+	AfterTealOp(cx *EvalContext, evalError error) error
 
-// LineToPC converts line to pc
-// Return 0 on unsuccess
-func (d *DebugState) LineToPC(line int) int {
-	if len(d.PCOffset) == 0 || line < 1 {
-		return 0
-	}
+	// BeforeInnerTxnGroup is called before an inner transaction group is executed.
+	// Each inner transaction within the group calls BeforeTxn and subsequent hooks, as described
+	// in the lifecycle diagram.
+	BeforeInnerTxnGroup(ep *EvalParams) error
 
-	lines := strings.Split(d.Disassembly, "\n")
-	if line > len(lines) {
-		return 0
-	}
-	offset := len(strings.Join(lines[:line], "\n"))
-
-	for i := 0; i < len(d.PCOffset); i++ {
-		if d.PCOffset[i].Offset >= offset {
-			return d.PCOffset[i].PC
-		}
-	}
-	return 0
+	// AfterInnerTxnGroup is called after an inner transaction group has been executed.
+	AfterInnerTxnGroup(ep *EvalParams) error
 }
 
-// PCToLine converts pc to line
-// Return 0 on unsuccess
-func (d *DebugState) PCToLine(pc int) int {
-	if len(d.PCOffset) == 0 {
-		return 0
-	}
-
-	offset := 0
-	for i := 0; i < len(d.PCOffset); i++ {
-		if d.PCOffset[i].PC >= pc {
-			offset = d.PCOffset[i].Offset
-			break
-		}
-	}
+// NullDebuggerHook implements DebuggerHook, but all of its hook methods do nothing
+type NullDebuggerHook struct{}
 
-	one := 1
-	// handle end of the program
-	if offset == 0 {
-		offset = d.PCOffset[len(d.PCOffset)-1].Offset
-		one = 0
-	}
-	if offset > len(d.Disassembly) {
-		return 0
-	}
-
-	return len(strings.Split(d.Disassembly[:offset], "\n")) - one
+// BeforeTxn does nothing
+func (null NullDebuggerHook) BeforeTxn(ep *EvalParams, groupIndex int) error {
+	return nil
```

Comment:
I recently learned this is legitimate for `go fmt` and I think is clearer for a string of similar, empty methods:

```
func (n NullDebuggerHook) BeforeTxn(ep *EvalParams, groupIndex int) error { return nil }
func (n NullDebuggerHook) AfterTxn(ep *EvalParams, groupIndex int) error    { return nil }
```

I continue to wonder about the `error` return here. I don't think `Tracers` ought to have the right to error.  They can observe, but it's not their place to report an error. Where can it go? If they want to not something for eventual reporting, they can stash that info inside their own state.

Reply:
I tried to get this to work, but our linter won't accept it because it wants every exported function to have a docstring.